core/optimization/src/main/java/it/unibz/inf/ontop/iq/optimizer/impl/DisjunctionOfEqualitiesMergingSimplifierImpl.java

```diff
@@ -0,0 +1,474 @@
+                return false;
+            var f = (ImmutableFunctionalTerm) target;
+
+            if(f.getFunctionSymbol() instanceof DBInFunctionSymbol)
+                return canMergeWith(f.getTerms(), in.getTerms());
+            if(isBooleanOperation(f))
+                return f.getTerms().stream()
+                        .anyMatch(child -> canMergeInto(child, in));
+            return false;
+        }
+
+        /**
+         * EXAMPLE SCENARIO: (x AND y) OR z
+         * where x, y are any expressions and z is an IN expression.
+         * Here, a disjunction with z must be applied to x and to y.
+         */
+        private ImmutableTerm mergeOppositeOperation(ImmutableFunctionalTerm operation, ImmutableFunctionalTerm in, boolean conjunction) {
+            var canCancel = operation.getTerms().stream()
+                    .collect(ImmutableCollectors.partitioningBy(
+                            child -> canMergeInto(child, in))
+                    );
+
+            var finalTerms = Streams.concat(
+                    canCancel.get(false).isEmpty()
+                            ? Stream.of()
+                            : Stream.of(termFactory.getImmutableExpression(
+                                    operatorSymbol(2, conjunction),
+                                    conjunctionDisjunctionOf(canCancel.get(false), !conjunction),
+                                    in
+                    )),
+                    canCancel.get(true).stream()
+                            .map(term -> mergeInto(term, in, conjunction))
+            ).collect(ImmutableCollectors.toList());
+
+            return conjunctionDisjunctionOf(finalTerms, !conjunction);
+        }
+
+        /**
+         * Determines whether the term `dependent` is implied by the term `determinant` in temrs of IN operations,
+         * where IN1 implies IN2 if they both share the same search term and the list of matches of IN1 is a subset
+         * of the list of matches of IN2.
+         */
+        private boolean isImpliedBy(ImmutableFunctionalTerm determinant, ImmutableFunctionalTerm dependent) {
+            if(determinant.getFunctionSymbol() instanceof DBInFunctionSymbol && dependent.getFunctionSymbol() instanceof DBInFunctionSymbol) {
+                //IN1 --> IN2 <==> search(IN1) = search(IN2) AND terms(IN1) subsetof terms(IN2)
+                return determinant.getTerm(0).equals(dependent.getTerm(0))
+                        && Sets.difference(ImmutableSet.copyOf(determinant.getTerms()), ImmutableSet.copyOf(dependent.getTerms())).isEmpty();
+            }
+            if(determinant.getFunctionSymbol() instanceof DBInFunctionSymbol && isBooleanOperation(dependent)) {
+                if(dependent.getFunctionSymbol() instanceof DBAndFunctionSymbol) {
+                    return dependent.getTerms().stream()
+                            .allMatch(term -> (term instanceof ImmutableFunctionalTerm) && isImpliedBy(determinant, (ImmutableFunctionalTerm) term));
```

Comment:
a second version of isImpliedBy could be applicable to ImmutableTerm, with the instanceof check and the type case to shorten these four lines

Reply:
I believe this should now be done, too, if I understood you correctly.